### PR TITLE
Support f-strings for common cases

### PR DESCRIPTION
**Pull request by ehsantn**
Provides support for common f-strings cases (without format specs). Essentially, adds support for `FORMAT_VALUE` and `BUILD_STRINGS` opcodes.
https://docs.python.org/3/reference/lexical_analysis.html#f-strings

### CODE REVIEW

**Comment by esc** [comment]
@ehsantn thank you for posting this, I have added it to the queue for review.

File: numba/core/interpreter.py [hunk 1]
```diff
@@ -467,6 +467,33 @@ def op_UNPACK_SEQUENCE(self, inst, iterable, stores, tupleobj):
                                           loc=self.loc)
             self.store(expr, st)
 
+    def op_FORMAT_VALUE(self, inst, value, res, strvar):
+        """
+        FORMAT_VALUE(flags): flags argument specifies format spec which is not supported
+        yet. Currently, we just call str() on the value.
```
Comment (stuartarchibald) [inline numba/core/interpreter.py:473]:
```suggestion
        yet. Currently, str() is simply called on the value.
```

File: numba/core/dataflow.py [hunk 2]
```diff
@@ -169,6 +169,37 @@ def op_UNPACK_SEQUENCE(self, info, inst):
         for st in reversed(stores):
             info.push(st)
 
+    def op_FORMAT_VALUE(self, info, inst):
+        """
+        FORMAT_VALUE(flags): flags argument specifies format spec which is
+        not supported yet. Currently, we just call str() on the value.
```
Comment (stuartarchibald) [inline numba/core/dataflow.py:175]:
```suggestion
        not supported yet. Currently, str() is simply called on the value.
```

File: numba/core/byteflow.py [hunk 3]
```diff
@@ -271,6 +271,37 @@ def dispatch(self, state):
...
+
+    def op_BUILD_STRING(self, state, inst):
+        """
+        BUILD_STRING(count): Concatenates count strings from the stack and
+        pushes the resulting string onto the stack.
+        Required for supporting f-strings.
+        https://docs.python.org/3/library/dis.html#opcode-BUILD_STRING
+        """
+        count = inst.arg
+        assert count > 0, "invalid BUILD_STRING count"
```
Comment (stuartarchibald) [inline numba/core/byteflow.py:299]: This:
```python
from numba import njit

@njit
def foo():
    return f""

print(foo())
```
whilst entirely contrived hits this, could it be handled?

File: numba/tests/test_unicode.py [hunk 4]
```diff
@@ -2549,6 +2549,20 @@ def foo(strinst):
...
+        def impl1(a):
+            return f"AA_{a+3}_B"
+
+        # does not require concatenation
+        def impl2(a):
+            return f"{a+2}"
+
+        self.assertEqual(impl1(3), njit(impl1)(3))
+        self.assertEqual(impl2(2), njit(impl2)(2))
+
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py:2565]: Suggest testing with string input and also with something for which Numba has no `str` support, which is just about anything other than ints and strings, so as to ensure something sensible appears as an error message.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the PR, great to see this regularly requested feature implemented. The implementation seems good in general, there's just a few minor things to resolve, most notably making sure that deviations from the Python spec are documented. Thanks again!

File: docs/source/reference/pysupported.rst [hunk 5]
```diff
@@ -312,6 +312,9 @@ supported:
 * ``.upper()``
 * ``.zfill()``
 
+Regular string literals (e.g. ``"ABC"``) as well as f-strings without format specs
+(e.g. ``"ABC_{a+1}"``) are supported in :term:`nopython mode`.
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply docs/source/reference/pysupported.rst]: Makes sense. Updated to reflect this.

File: numba/core/byteflow.py [hunk 6]
```diff
@@ -271,6 +271,37 @@ def dispatch(self, state):
...
+        """
+        FORMAT_VALUE(flags): flags argument specifies format spec which is
+        not supported yet. Currently, we just call str() on the value.
+        Pops a value from stack and pushes results back.
+        Required for supporting f-strings.
+        https://docs.python.org/3/library/dis.html#opcode-FORMAT_VALUE
+        """
+        if inst.arg != 0:
+            msg = "format spec in f-strings not supported yet"
+            raise UnsupportedError(msg, loc=self.get_debug_loc(inst.lineno))
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/core/byteflow.py:284]: Added.

File: numba/core/dataflow.py [hunk 7]
```diff
@@ -169,6 +169,37 @@ def op_UNPACK_SEQUENCE(self, info, inst):
...
+        Pops a value from stack and pushes results back.
+        Required for supporting f-strings.
+        https://docs.python.org/3/library/dis.html#opcode-FORMAT_VALUE
+        """
+        if inst.arg != 0:
+            msg = "format spec in f-strings not supported yet"
+            raise UnsupportedError(msg, loc=self.get_debug_loc(inst.lineno))
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/core/dataflow.py:182]: Added.

File: numba/core/dataflow.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/core/dataflow.py]: Done.

File: numba/core/byteflow.py (hunk 3 above)
Comment (ehsantn) [reply numba/core/byteflow.py]: Added support for it.

File: numba/tests/test_unicode.py (hunk 4 above)
Comment (ehsantn) [reply numba/tests/test_unicode.py:2590]: Done.

**Comment by ehsantn** [comment]
Thanks for the review @stuartarchibald . I applied the changes. Let me know if anything else is needed.

**Comment by stuartarchibald** [comment]
Thanks for the fixes @ehsantn, there's a couple of test cases failing, coming from `dataflow.py`, think it is because there's no debug location information available in the older Python 3.6 analyser. There's also some `flake8` failures.

**Comment by ehsantn** [comment]
Sorry for the delay. CI is passing now.

**Comment by stuartarchibald** [comment]
> Sorry for the delay. CI is passing now.

Thanks @ehsantn

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes!

**Pull request merged**